src/Storefront/Resources/app/storefront/src/plugin-system/plugin.manager.js

```diff
@@ -266,18 +266,65 @@ class PluginManagerSingleton {
         }
 
         if (typeof selector === 'string') {
-            selector = document.querySelectorAll(selector);
+            selector = PluginManagerSingleton._queryElements(selector);
         }
 
         return Iterator.iterate(selector, el => {
             PluginManagerSingleton._initializePluginOnElement(el, pluginClass, options, pluginName);
         });
     }
 
+    /**
+     * Determs the way to query the elements.
+     *
+     * [data-*] => querySelectorAll
+     * #fooBar => getElementById
+     * #foo_bar => getElementById
+     * #foo-bar => getElementById
+     * #foo .bar => querySelectorAll
+     * .fooBar => getElementsByClassName
+     * .FOO => getElementsByClassName
+     * .foo_bar => getElementsByClassName
+     * .foo-bar => getElementsByClassName
+     * .foo .bar => querySelectorAll
+     *
+     * FOO => getElementsByTagName
+     * FOO .bar => querySelectorAll
+     *
+     * For performance reason used regex based on common characters `a-zA-Z1-9_-`
+     * instead of the entire compatible characters
+     *
+     * @param {string} selector
+     *
+     * @return {NodeList|HTMLCollection|Array}
+     */
+    static _queryElements(selector) {
+        // While the use of attributes is the most common right now, we should return early.
+        if (selector.startsWith('[')) {
```

Comment:
I don't think we need this fallback right here at the beginning. The string operations in your `if`s are evaluated much quicker than the `document.*` methods.

Reply:
It isn't a fallback, is it? 🤔

with this early-return, I want to get rid of any further useless checks, especially the [regex for tags](https://github.com/shopware/platform/pull/2265/files#diff-c0dd49c71eb6bb27d09d53dabef68e9dd2f243aee3ec98d102d3f727d67bcfb6R317)

Reply:
It does the same as if none of the if statements evaluate to true.

Compared with the DOM operations that will be called the if statements won't have a measurable impact on performance. Even the regex won't.